util/reggen/reg_block.py

```diff
@@ -190,34 +207,18 @@ def _validate_async(self, name: Optional[str]) -> None:
             async designation of the bus interface.
             ''')
 
-    def handle_async(self, name: Optional[str], clocks: Clocking, bus: BusInterfaces) -> None:
+        if name:
+            self.clocks[name] = clk
 
-        reg_clocks = {}
-        if bus.device_async.get(name) is not None:
-            cname = bus.device_async[name]
-            self.async_if = True
-            reg_clocks[cname] = clocks.get_by_clock(cname)
-        else:
-
-            for r in self.registers:
-                clock = r.handle_async(clocks)
-                if clock:
-                    reg_clocks[r.async_name] = clock
-
-            for mr in self.multiregs:
-                clock = mr.handle_async(clocks)
-                if clock:
-                    reg_clocks[r.async_name] = clock
+    def _handle_register(self, where: str, body: object, clocks: Clocking) -> None:
+        reg = Register.from_raw(self._reg_width,
+                                self.offset, self._params, body, clocks)
 
-        self.clocks = reg_clocks
-        self._validate_async(name)
+        self._validate_async(reg.async_name, reg.async_clk)
 
-    def _handle_register(self, where: str, body: object) -> None:
-        reg = Register.from_raw(self._reg_width,
-                                self.offset, self._params, body)
         self.add_register(reg)
 
-    def _handle_reserved(self, where: str, body: object) -> None:
+    def _handle_reserved(self, where: str, body: object, clocks: Optional[Clocking]) -> None:
```

Comment:
@rswarbrick this part is kind of annoying.
I originally followed some examples online to pickout variable number of arguments to supply. 

However, `mypy` just kept complaining about function and callable typing issues. In the end I gave up and just made it optional on everyone.

Reply:
Well, the point is that `_handle_reserved` is put in a dictionary of "handlers", which are all called with the same arguments. So MyPy is actually complaining correctly: if we removed the `clocks` argument then the call at the end of `add_raw` would fail at runtime. I don't think the private method is called in any other context, so I'd suggest just making the type signature match the other handlers.

Reply:
i should have clarified i was doing [this](https://stackoverflow.com/questions/50496821/how-to-call-functions-in-a-dict-that-take-different-numbers-of-arguments).  And it actually worked just fine, but mypy just didn't like the typing. 

The main thing it was complaining about (which i could not figure out) is that the input argument was a function and not a callable.  That was...odd to me, because I thought all functions were callables.

Reply:
I suspect that it was complaining that it wasn't a callable with the expected arguments. The point is that MyPy types `def foo(a: int) -> object` as `Callable[[int], object]` and types `def bar(a: int, b: str) -> object` as `Callable[[int, str], object]` and these are different types.

You **could** do something like add an explicit type to the definition of the list, which might work (`Union[Callable[[int], object], Callable[[int, str], object]]`), but it's probably a sign that we're doing something unnecessarily complicated.